{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6fe20f39",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import collections\n",
    "import seaborn as sns\n",
    "import matplotlib \n",
    "import matplotlib.pyplot as plt\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fd6215a",
   "metadata": {},
   "source": [
    "## Get barcodes for each clone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b476b06",
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = \"YF-2818_8HetClonesHashtagMix_hLLL_multiome\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e0c672f",
   "metadata": {},
   "source": [
    "We use HTO output and save each clone barcode in a separate file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5dc296fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "hto_adata_path = f'/data/peer/landm/Projects/sail-dogma/yi-clone-samples/data/{SAMPLE}/HTO/adataFinal_YF-2818_8HetClonesHashtagMix_hLLL_multiome.h5ad'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8abeeaed",
   "metadata": {},
   "outputs": [],
   "source": [
    "hto_adata = sc.read_h5ad(hto_adata_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "449e737a",
   "metadata": {},
   "outputs": [],
   "source": [
    "hto_adata.obs['cr_barcodes'] = hto_adata.obs['barcode_sequence'] + '-1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "377ed228",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unmapped</th>\n",
       "      <th>barcode_sequence</th>\n",
       "      <th>hash_id</th>\n",
       "      <th>cr_barcodes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcodes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AGCGCTCAGAACCTAC</th>\n",
       "      <td>91</td>\n",
       "      <td>AGCGCTCAGAACCTAC</td>\n",
       "      <td>Doublet</td>\n",
       "      <td>AGCGCTCAGAACCTAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGTCAACAACAGCCT</th>\n",
       "      <td>112</td>\n",
       "      <td>GGGTCAACAACAGCCT</td>\n",
       "      <td>A0258</td>\n",
       "      <td>GGGTCAACAACAGCCT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGGAACCAGCATTAT</th>\n",
       "      <td>37</td>\n",
       "      <td>GCGGAACCAGCATTAT</td>\n",
       "      <td>A0257</td>\n",
       "      <td>GCGGAACCAGCATTAT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGCTAAAGTTGTCAA</th>\n",
       "      <td>83</td>\n",
       "      <td>AGGCTAAAGTTGTCAA</td>\n",
       "      <td>A0255</td>\n",
       "      <td>AGGCTAAAGTTGTCAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATTAGTGTACGATTG</th>\n",
       "      <td>30</td>\n",
       "      <td>GATTAGTGTACGATTG</td>\n",
       "      <td>A0254</td>\n",
       "      <td>GATTAGTGTACGATTG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCCTCACCAATAAGCA</th>\n",
       "      <td>182</td>\n",
       "      <td>CCCTCACCAATAAGCA</td>\n",
       "      <td>A0253</td>\n",
       "      <td>CCCTCACCAATAAGCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCCTTAGAGTGAAGTG</th>\n",
       "      <td>113</td>\n",
       "      <td>GCCTTAGAGTGAAGTG</td>\n",
       "      <td>A0253</td>\n",
       "      <td>GCCTTAGAGTGAAGTG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TATCCGGAGCCGCAAC</th>\n",
       "      <td>162</td>\n",
       "      <td>TATCCGGAGCCGCAAC</td>\n",
       "      <td>A0258</td>\n",
       "      <td>TATCCGGAGCCGCAAC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAGGGTGAGAACCTGT</th>\n",
       "      <td>83</td>\n",
       "      <td>CAGGGTGAGAACCTGT</td>\n",
       "      <td>Doublet</td>\n",
       "      <td>CAGGGTGAGAACCTGT-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATTTGCAGCATGTCG</th>\n",
       "      <td>64</td>\n",
       "      <td>GATTTGCAGCATGTCG</td>\n",
       "      <td>A0256</td>\n",
       "      <td>GATTTGCAGCATGTCG-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16150 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  unmapped  barcode_sequence  hash_id         cr_barcodes\n",
       "cell_barcodes                                                            \n",
       "AGCGCTCAGAACCTAC        91  AGCGCTCAGAACCTAC  Doublet  AGCGCTCAGAACCTAC-1\n",
       "GGGTCAACAACAGCCT       112  GGGTCAACAACAGCCT    A0258  GGGTCAACAACAGCCT-1\n",
       "GCGGAACCAGCATTAT        37  GCGGAACCAGCATTAT    A0257  GCGGAACCAGCATTAT-1\n",
       "AGGCTAAAGTTGTCAA        83  AGGCTAAAGTTGTCAA    A0255  AGGCTAAAGTTGTCAA-1\n",
       "GATTAGTGTACGATTG        30  GATTAGTGTACGATTG    A0254  GATTAGTGTACGATTG-1\n",
       "...                    ...               ...      ...                 ...\n",
       "CCCTCACCAATAAGCA       182  CCCTCACCAATAAGCA    A0253  CCCTCACCAATAAGCA-1\n",
       "GCCTTAGAGTGAAGTG       113  GCCTTAGAGTGAAGTG    A0253  GCCTTAGAGTGAAGTG-1\n",
       "TATCCGGAGCCGCAAC       162  TATCCGGAGCCGCAAC    A0258  TATCCGGAGCCGCAAC-1\n",
       "CAGGGTGAGAACCTGT        83  CAGGGTGAGAACCTGT  Doublet  CAGGGTGAGAACCTGT-1\n",
       "GATTTGCAGCATGTCG        64  GATTTGCAGCATGTCG    A0256  GATTTGCAGCATGTCG-1\n",
       "\n",
       "[16150 rows x 4 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hto_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "39f571f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "hto_barcodes = hto_adata.obs['cr_barcodes'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fb4ebf21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='Hash Id', ylabel='count'>"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkQAAAGwCAYAAABIC3rIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAxjUlEQVR4nO3de3RU5b3/8c8EmJAgCQTITUMEbLkZQLHGaA1BMCFQqpXjhYtgzQHLCVqJB3PSH2LAVhQQoR6q1YrUYyheqtgiRS4CoRBQwBhBRMDY0JIEl5AMBHN/fn+0mTIm3EKSmfC8X2vttWY/zzN7f7+Ki49775lxGGOMAAAALObn7QIAAAC8jUAEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGC9tt4uoDWora3VkSNH1LFjRzkcDm+XAwAAzoMxRidOnFBkZKT8/M5+DYhAdB6OHDmiqKgob5cBAAAa4fDhw7riiivOuoZAdB46duwo6Z//QIOCgrxcDQAAOB8ul0tRUVHuv8fPhkB0HupukwUFBRGIAABoZc7ncRceqgYAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYr623CwAAX/a/j/zZ2yVctGnPjPZ2CYDP4woRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1+Ng9cIFueu4mb5dw0bY+uNXbJQCATyEQAQBguczMTG+XcNEutgdumQEAAOsRiAAAgPUIRAAAwHo8QwTgvGyOH+LtEi7akOzN3i4BgI/iChEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOt5NRBlZ2dr9OjRioyMlMPh0MqVKz3mHQ5Hg9v8+fPda6688sp680899ZTHcfLy8nTzzTerffv2ioqK0rx581qiPQAA0Ep4NRCVlZVp4MCBWrJkSYPzhYWFHtvSpUvlcDg0ZswYj3Vz5szxWPfggw+651wulxITExUdHa1du3Zp/vz5yszM1IsvvtisvQEAgNbDq99DlJycrOTk5DPOh4eHe+y/++67Gjp0qHr27Okx3rFjx3pr62RlZamyslJLly6V0+lU//79lZubq4ULF2rKlCkNvqeiokIVFRXufZfLdb4tAQCAVqjVPENUXFys9957TykpKfXmnnrqKXXp0kXXXHON5s+fr+rqavdcTk6O4uPj5XQ63WNJSUnav3+/jh8/3uC55s6dq+DgYPcWFRXV9A0BAACf0WoC0e9//3t17NhRd9xxh8f4Qw89pBUrVmjjxo164IEH9OSTT+rRRx91zxcVFSksLMzjPXX7RUVFDZ4rIyNDpaWl7u3w4cNN3A0AAPAlreanO5YuXarx48erffv2HuNpaWnu1wMGDJDT6dQDDzyguXPnyt/fv1Hn8vf3b/R7AeBS8KsJ/+HtEi7a/3vtLW+XgFakVVwh2rJli/bv36///M//POfa2NhYVVdX66uvvpL0z+eQiouLPdbU7Z/puSMAAGCXVhGIXn75ZQ0ePFgDBw4859rc3Fz5+fkpNDRUkhQXF6fs7GxVVVW516xbt069e/dW586dm61mAADQeng1EJ08eVK5ubnKzc2VJOXn5ys3N1cFBQXuNS6XS2+++WaDV4dycnK0aNEiffLJJ/ryyy+VlZWl6dOna8KECe6wM27cODmdTqWkpGjv3r16/fXXtXjxYo9bbQAAwG5efYZo586dGjp0qHu/LqRMmjRJy5YtkyStWLFCxhiNHTu23vv9/f21YsUKZWZmqqKiQj169ND06dM9wk5wcLDWrl2r1NRUDR48WF27dtWsWbPO+JF7AABgH68GooSEBBljzrpmypQpZwwv1157rbZv337O8wwYMEBbtmxpVI0AAODS12o+ZQYAQHPb96sPvF3CRev7/27xdgmtUqt4qBoAAKA5EYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWa+vtAi4Fg2e86u0SLtqu+RO9XQIAAF7DFSIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsJ5XA1F2drZGjx6tyMhIORwOrVy50mP+vvvuk8Ph8NhGjBjhsebYsWMaP368goKC1KlTJ6WkpOjkyZMea/Ly8nTzzTerffv2ioqK0rx585q7NQAA0Ip4NRCVlZVp4MCBWrJkyRnXjBgxQoWFhe7tD3/4g8f8+PHjtXfvXq1bt06rVq1Sdna2pkyZ4p53uVxKTExUdHS0du3apfnz5yszM1Mvvvhis/UFAABaF6/+uGtycrKSk5PPusbf31/h4eENzu3bt09r1qzRRx99pOuuu06S9Nxzz2nkyJFasGCBIiMjlZWVpcrKSi1dulROp1P9+/dXbm6uFi5c6BGcTldRUaGKigr3vsvlamSHAACgNfD5Z4g2bdqk0NBQ9e7dW1OnTtU333zjnsvJyVGnTp3cYUiShg8fLj8/P+3YscO9Jj4+Xk6n070mKSlJ+/fv1/Hjxxs859y5cxUcHOzeoqKimqk7AADgC3w6EI0YMUKvvvqqNmzYoKefflqbN29WcnKyampqJElFRUUKDQ31eE/btm0VEhKioqIi95qwsDCPNXX7dWu+KyMjQ6Wlpe7t8OHDTd0aAADwIV69ZXYu99xzj/t1TEyMBgwYoF69emnTpk0aNmxYs53X399f/v7+zXZ8AADgW3z6CtF39ezZU127dtXBgwclSeHh4Tp69KjHmurqah07dsz93FF4eLiKi4s91tTtn+nZJAAAYJdWFYj+/ve/65tvvlFERIQkKS4uTiUlJdq1a5d7zQcffKDa2lrFxsa612RnZ6uqqsq9Zt26derdu7c6d+7csg0AAACf5NVAdPLkSeXm5io3N1eSlJ+fr9zcXBUUFOjkyZOaMWOGtm/frq+++kobNmzQbbfdpquuukpJSUmSpL59+2rEiBGaPHmyPvzwQ23dulXTpk3TPffco8jISEnSuHHj5HQ6lZKSor179+r111/X4sWLlZaW5q22AQCAj/FqINq5c6euueYaXXPNNZKktLQ0XXPNNZo1a5batGmjvLw8/fjHP9b3v/99paSkaPDgwdqyZYvH8z1ZWVnq06ePhg0bppEjR+qHP/yhx3cMBQcHa+3atcrPz9fgwYP1yCOPaNasWWf8yD0AALCPVx+qTkhIkDHmjPPvv//+OY8REhKi5cuXn3XNgAEDtGXLlguuDwAA2KFVPUMEAADQHAhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6Xv3pDrRuBXNivF3CRes+61NvlwAA8AFcIQIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANbzaiDKzs7W6NGjFRkZKYfDoZUrV7rnqqqqlJ6erpiYGHXo0EGRkZGaOHGijhw54nGMK6+8Ug6Hw2N76qmnPNbk5eXp5ptvVvv27RUVFaV58+a1RHsAAKCV8GogKisr08CBA7VkyZJ6c6dOndLu3bv12GOPaffu3Xr77be1f/9+/fjHP663ds6cOSosLHRvDz74oHvO5XIpMTFR0dHR2rVrl+bPn6/MzEy9+OKLzdobAABoPdp68+TJyclKTk5ucC44OFjr1q3zGPvf//1fXX/99SooKFD37t3d4x07dlR4eHiDx8nKylJlZaWWLl0qp9Op/v37Kzc3VwsXLtSUKVOarhkAANBqtapniEpLS+VwONSpUyeP8aeeekpdunTRNddco/nz56u6uto9l5OTo/j4eDmdTvdYUlKS9u/fr+PHjzd4noqKCrlcLo8NAABcurx6hehClJeXKz09XWPHjlVQUJB7/KGHHtK1116rkJAQbdu2TRkZGSosLNTChQslSUVFRerRo4fHscLCwtxznTt3rneuuXPnavbs2c3YDQAA8CWtIhBVVVXprrvukjFGzz//vMdcWlqa+/WAAQPkdDr1wAMPaO7cufL392/U+TIyMjyO63K5FBUV1bjiAQCAz/P5QFQXhv72t7/pgw8+8Lg61JDY2FhVV1frq6++Uu/evRUeHq7i4mKPNXX7Z3ruyN/fv9FhCgAAtD4+/QxRXRg6cOCA1q9fry5dupzzPbm5ufLz81NoaKgkKS4uTtnZ2aqqqnKvWbdunXr37t3g7TIAAGAfr14hOnnypA4ePOjez8/PV25urkJCQhQREaH/+I//0O7du7Vq1SrV1NSoqKhIkhQSEiKn06mcnBzt2LFDQ4cOVceOHZWTk6Pp06drwoQJ7rAzbtw4zZ49WykpKUpPT9eePXu0ePFiPfvss17pGQAA+B6vBqKdO3dq6NCh7v2653YmTZqkzMxM/elPf5IkDRo0yON9GzduVEJCgvz9/bVixQplZmaqoqJCPXr00PTp0z2e/wkODtbatWuVmpqqwYMHq2vXrpo1axYfuQcAAG5eDUQJCQkyxpxx/mxzknTttddq+/bt5zzPgAEDtGXLlguuDwAA2MGnnyECAABoCQQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwXqMC0S233KKSkpJ64y6XS7fccsvF1gQAANCiGhWINm3apMrKynrj5eXl2rJly0UXBQAA0JLaXsjivLw89+vPPvtMRUVF7v2amhqtWbNGl19+edNVBwAA0AIuKBANGjRIDodDDoejwVtjAQEBeu6555qsOAAAgJZwQYEoPz9fxhj17NlTH374obp16+aeczqdCg0NVZs2bZq8SAAAgOZ0QYEoOjpaklRbW9ssxQAAAHjDBQWi0x04cEAbN27U0aNH6wWkWbNmXXRhAAAALaVRgeill17S1KlT1bVrV4WHh8vhcLjnHA4HgQgAALQqjQpEv/zlL/WrX/1K6enpTV0PAABAi2vU9xAdP35cd95550WfPDs7W6NHj1ZkZKQcDodWrlzpMW+M0axZsxQREaGAgAANHz5cBw4c8Fhz7NgxjR8/XkFBQerUqZNSUlJ08uRJjzV5eXm6+eab1b59e0VFRWnevHkXXTsAALh0NCoQ3XnnnVq7du1Fn7ysrEwDBw7UkiVLGpyfN2+efv3rX+uFF17Qjh071KFDByUlJam8vNy9Zvz48dq7d6/WrVunVatWKTs7W1OmTHHPu1wuJSYmKjo6Wrt27dL8+fOVmZmpF1988aLrBwAAl4ZG3TK76qqr9Nhjj2n79u2KiYlRu3btPOYfeuih8zpOcnKykpOTG5wzxmjRokWaOXOmbrvtNknSq6++qrCwMK1cuVL33HOP9u3bpzVr1uijjz7SddddJ0l67rnnNHLkSC1YsECRkZHKyspSZWWlli5dKqfTqf79+ys3N1cLFy70CE4AAMBejQpEL774oi677DJt3rxZmzdv9phzOBznHYjOJj8/X0VFRRo+fLh7LDg4WLGxscrJydE999yjnJwcderUyR2GJGn48OHy8/PTjh079JOf/EQ5OTmKj4+X0+l0r0lKStLTTz+t48ePq3PnzvXOXVFRoYqKCve+y+W66H4AAIDvalQgys/Pb+o66qn7WZCwsDCP8bCwMPdcUVGRQkNDPebbtm2rkJAQjzU9evSod4y6uYYC0dy5czV79uymaQQAAPi8Rj1DdKnLyMhQaWmpezt8+LC3SwIAAM2oUVeI7r///rPOL126tFHFnC48PFySVFxcrIiICPd4cXGxBg0a5F5z9OhRj/dVV1fr2LFj7veHh4eruLjYY03dft2a7/L395e/v/9F9wAAAFqHRn/s/vTt6NGj+uCDD/T222+rpKSkSQrr0aOHwsPDtWHDBveYy+XSjh07FBcXJ0mKi4tTSUmJdu3a5V7zwQcfqLa2VrGxse412dnZqqqqcq9Zt26devfu3eDtMgAAYJ9GXSF655136o3V1tZq6tSp6tWr13kf5+TJkzp48KB7Pz8/X7m5uQoJCVH37t318MMP65e//KW+973vqUePHnrssccUGRmp22+/XZLUt29fjRgxQpMnT9YLL7ygqqoqTZs2Tffcc48iIyMlSePGjdPs2bOVkpKi9PR07dmzR4sXL9azzz7bmNYBAMAlqNG/ZfZdfn5+SktLU0JCgh599NHzes/OnTs1dOhQ935aWpokadKkSVq2bJkeffRRlZWVacqUKSopKdEPf/hDrVmzRu3bt3e/JysrS9OmTdOwYcPk5+enMWPG6Ne//rV7Pjg4WGvXrlVqaqoGDx6srl27atasWXzkHgAAuDVZIJKkQ4cOqbq6+rzXJyQkyBhzxnmHw6E5c+Zozpw5Z1wTEhKi5cuXn/U8AwYM0JYtW867LgAAYJdGBaK6Kzl1jDEqLCzUe++9p0mTJjVJYQAAAC2lUYHo448/9tj38/NTt27d9Mwzz5zzE2gAAAC+plGBaOPGjU1dBwAAgNdc1DNEX3/9tfbv3y9J6t27t7p169YkRQEAALSkRn0PUVlZme6//35FREQoPj5e8fHxioyMVEpKik6dOtXUNQIAADSrRgWitLQ0bd68WX/+859VUlKikpISvfvuu9q8ebMeeeSRpq4RAACgWTXqltkf//hHvfXWW0pISHCPjRw5UgEBAbrrrrv0/PPPN1V9AAAAza5RV4hOnTpV71foJSk0NJRbZgAAoNVpVCCKi4vT448/rvLycvfYt99+q9mzZ7t/ZwwAAKC1aNQts0WLFmnEiBG64oorNHDgQEnSJ598In9/f61du7ZJCwQAAGhujQpEMTExOnDggLKysvT5559LksaOHavx48crICCgSQsEAABobo0KRHPnzlVYWJgmT57sMb506VJ9/fXXSk9Pb5LiAAAAWkKjniH67W9/qz59+tQb79+/v1544YWLLgoAAKAlNSoQFRUVKSIiot54t27dVFhYeNFFAQAAtKRGBaKoqCht3bq13vjWrVsVGRl50UUBAAC0pEY9QzR58mQ9/PDDqqqq0i233CJJ2rBhgx599FG+qRoAALQ6jQpEM2bM0DfffKP/+q//UmVlpSSpffv2Sk9PV0ZGRpMWCAAA0NwaFYgcDoeefvppPfbYY9q3b58CAgL0ve99T/7+/k1dHwAAQLNrVCCqc9lll+kHP/hBU9UCAADgFY16qBoAAOBSQiACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArOfzgejKK6+Uw+Got6WmpkqSEhIS6s397Gc/8zhGQUGBRo0apcDAQIWGhmrGjBmqrq72RjsAAMAHtfV2Aefy0Ucfqaamxr2/Z88e3XrrrbrzzjvdY5MnT9acOXPc+4GBge7XNTU1GjVqlMLDw7Vt2zYVFhZq4sSJateunZ588smWaQIAAPg0nw9E3bp189h/6qmn1KtXLw0ZMsQ9FhgYqPDw8Abfv3btWn322Wdav369wsLCNGjQID3xxBNKT09XZmamnE5ns9YPAAB8n8/fMjtdZWWlXnvtNd1///1yOBzu8aysLHXt2lVXX321MjIydOrUKfdcTk6OYmJiFBYW5h5LSkqSy+XS3r17GzxPRUWFXC6XxwYAAC5dPn+F6HQrV65USUmJ7rvvPvfYuHHjFB0drcjISOXl5Sk9PV379+/X22+/LUkqKiryCEOS3PtFRUUNnmfu3LmaPXt28zQBAAB8TqsKRC+//LKSk5MVGRnpHpsyZYr7dUxMjCIiIjRs2DAdOnRIvXr1atR5MjIylJaW5t53uVyKiopqfOEAAMCntZpA9Le//U3r1693X/k5k9jYWEnSwYMH1atXL4WHh+vDDz/0WFNcXCxJZ3zuyN/fX/7+/k1QNQAAaA1azTNEr7zyikJDQzVq1KizrsvNzZUkRURESJLi4uL06aef6ujRo+4169atU1BQkPr169ds9QIAgNajVVwhqq2t1SuvvKJJkyapbdt/l3zo0CEtX75cI0eOVJcuXZSXl6fp06crPj5eAwYMkCQlJiaqX79+uvfeezVv3jwVFRVp5syZSk1N5SoQAACQ1EoC0fr161VQUKD777/fY9zpdGr9+vVatGiRysrKFBUVpTFjxmjmzJnuNW3atNGqVas0depUxcXFqUOHDpo0aZLH9xYBAAC7tYpAlJiYKGNMvfGoqCht3rz5nO+Pjo7W6tWrm6M0AABwCWg1zxABAAA0FwIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANbz6UCUmZkph8PhsfXp08c9X15ertTUVHXp0kWXXXaZxowZo+LiYo9jFBQUaNSoUQoMDFRoaKhmzJih6urqlm4FAAD4sLbeLuBc+vfvr/Xr17v327b9d8nTp0/Xe++9pzfffFPBwcGaNm2a7rjjDm3dulWSVFNTo1GjRik8PFzbtm1TYWGhJk6cqHbt2unJJ59s8V4AAIBv8vlA1LZtW4WHh9cbLy0t1csvv6zly5frlltukSS98sor6tu3r7Zv364bbrhBa9eu1Weffab169crLCxMgwYN0hNPPKH09HRlZmbK6XQ2eM6KigpVVFS4910uV/M0BwAAfIJP3zKTpAMHDigyMlI9e/bU+PHjVVBQIEnatWuXqqqqNHz4cPfaPn36qHv37srJyZEk5eTkKCYmRmFhYe41SUlJcrlc2rt37xnPOXfuXAUHB7u3qKioZuoOAAD4Ap8ORLGxsVq2bJnWrFmj559/Xvn5+br55pt14sQJFRUVyel0qlOnTh7vCQsLU1FRkSSpqKjIIwzVzdfNnUlGRoZKS0vd2+HDh5u2MQAA4FN8+pZZcnKy+/WAAQMUGxur6OhovfHGGwoICGi28/r7+8vf37/Zjg8AAHyLT18h+q5OnTrp+9//vg4ePKjw8HBVVlaqpKTEY01xcbH7maPw8PB6nzqr22/ouSQAAGCnVhWITp48qUOHDikiIkKDBw9Wu3bttGHDBvf8/v37VVBQoLi4OElSXFycPv30Ux09etS9Zt26dQoKClK/fv1avH4AAOCbfPqW2X//939r9OjRio6O1pEjR/T444+rTZs2Gjt2rIKDg5WSkqK0tDSFhIQoKChIDz74oOLi4nTDDTdIkhITE9WvXz/de++9mjdvnoqKijRz5kylpqZySwwAALj5dCD6+9//rrFjx+qbb75Rt27d9MMf/lDbt29Xt27dJEnPPvus/Pz8NGbMGFVUVCgpKUm/+c1v3O9v06aNVq1apalTpyouLk4dOnTQpEmTNGfOHG+1BAAAfJBPB6IVK1acdb59+/ZasmSJlixZcsY10dHRWr16dVOXBgAALiGt6hkiAACA5kAgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1fDoQzZ07Vz/4wQ/UsWNHhYaG6vbbb9f+/fs91iQkJMjhcHhsP/vZzzzWFBQUaNSoUQoMDFRoaKhmzJih6urqlmwFAAD4sLbeLuBsNm/erNTUVP3gBz9QdXW1fvGLXygxMVGfffaZOnTo4F43efJkzZkzx70fGBjofl1TU6NRo0YpPDxc27ZtU2FhoSZOnKh27drpySefbNF+AACAb/LpQLRmzRqP/WXLlik0NFS7du1SfHy8ezwwMFDh4eENHmPt2rX67LPPtH79eoWFhWnQoEF64oknlJ6erszMTDmdzmbtAQAA+D6fvmX2XaWlpZKkkJAQj/GsrCx17dpVV199tTIyMnTq1Cn3XE5OjmJiYhQWFuYeS0pKksvl0t69exs8T0VFhVwul8cGAAAuXT59heh0tbW1evjhh3XTTTfp6quvdo+PGzdO0dHRioyMVF5entLT07V//369/fbbkqSioiKPMCTJvV9UVNTguebOnavZs2c3UycAAMDXtJpAlJqaqj179uivf/2rx/iUKVPcr2NiYhQREaFhw4bp0KFD6tWrV6POlZGRobS0NPe+y+VSVFRU4woHAAA+r1XcMps2bZpWrVqljRs36oorrjjr2tjYWEnSwYMHJUnh4eEqLi72WFO3f6bnjvz9/RUUFOSxAQCAS5dPByJjjKZNm6Z33nlHH3zwgXr06HHO9+Tm5kqSIiIiJElxcXH69NNPdfToUfeadevWKSgoSP369WuWugEAQOvi07fMUlNTtXz5cr377rvq2LGj+5mf4OBgBQQE6NChQ1q+fLlGjhypLl26KC8vT9OnT1d8fLwGDBggSUpMTFS/fv107733at68eSoqKtLMmTOVmpoqf39/b7YHAAB8hE9fIXr++edVWlqqhIQERUREuLfXX39dkuR0OrV+/XolJiaqT58+euSRRzRmzBj9+c9/dh+jTZs2WrVqldq0aaO4uDhNmDBBEydO9PjeIgAAYDefvkJkjDnrfFRUlDZv3nzO40RHR2v16tVNVRYAALjE+PQVIgAAgJZAIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9awKREuWLNGVV16p9u3bKzY2Vh9++KG3SwIAAD7AmkD0+uuvKy0tTY8//rh2796tgQMHKikpSUePHvV2aQAAwMusCUQLFy7U5MmT9dOf/lT9+vXTCy+8oMDAQC1dutTbpQEAAC9r6+0CWkJlZaV27dqljIwM95ifn5+GDx+unJyceusrKipUUVHh3i8tLZUkuVyuBo9fU/FtE1fc8s7U29mcKK9phkpaVmP6rv62uhkqaVmN6bus2s6+v6041QyVtKzG9F1eVdUMlbSsxvR9srysGSppWY3p+/S/81qrhvquGzPGnPsAxgL/+Mc/jCSzbds2j/EZM2aY66+/vt76xx9/3EhiY2NjY2NjuwS2w4cPnzMrWHGF6EJlZGQoLS3NvV9bW6tjx46pS5cucjgcLVqLy+VSVFSUDh8+rKCgoBY9tzfRN33bgL7p2wbe7NsYoxMnTigyMvKca60IRF27dlWbNm1UXFzsMV5cXKzw8PB66/39/eXv7+8x1qlTp+Ys8ZyCgoKs+g+oDn3bhb7tQt928VbfwcHB57XOioeqnU6nBg8erA0bNrjHamtrtWHDBsXFxXmxMgAA4AusuEIkSWlpaZo0aZKuu+46XX/99Vq0aJHKysr005/+1NulAQAAL7MmEN199936+uuvNWvWLBUVFWnQoEFas2aNwsLCvF3aWfn7++vxxx+vdwvvUkff9G0D+qZvG7SWvh3GnM9n0QAAAC5dVjxDBAAAcDYEIgAAYD0CEQAAsB6BCAAAWI9A1ExycnLUpk0bjRo1qt5cQUGBRo0apcDAQIWGhmrGjBmqPu13ot5++23deuut6tatm4KCghQXF6f333/f4xiZmZlyOBweW58+fTzWFBUV6d5771V4eLg6dOiga6+9Vn/84x+bp+F/8XbfX331Vb35uu3NN9+8ZPuWpISEhHprfvaznzVPw//iC33XMcYoOTlZDodDK1eubNI+v8vbfR87dkwPPvigevfurYCAAHXv3l0PPfSQ+3cXm4u3+5akBx54QL169VJAQIC6deum2267TZ9//nnzNPwvvtD3iy++qISEBAUFBcnhcKikpKRZej2dL/RdV8ctt9yiDh06KCgoSPHx8fr226b/DVECUTN5+eWX9eCDDyo7O1tHjhxxj9fU1GjUqFGqrKzUtm3b9Pvf/17Lli3TrFmz3Guys7N16623avXq1dq1a5eGDh2q0aNH6+OPP/Y4R//+/VVYWOje/vrXv3rMT5w4Ufv379ef/vQnffrpp7rjjjt011131TvOpdR3VFSUx1xhYaFmz56tyy67TMnJyZds33UmT57ssWbevHnN1rPkO31L0qJFi1rsp3W83feRI0d05MgRLViwQHv27NGyZcu0Zs0apaSkXNJ9S9LgwYP1yiuvaN++fXr//fdljFFiYqJqaprvx6Z9oe9Tp05pxIgR+sUvftFsfX6XL/Sdk5OjESNGKDExUR9++KE++ugjTZs2TX5+zRBfmuLHU+HpxIkT5rLLLjOff/65ufvuu82vfvUr99zq1auNn5+fKSoqco89//zzJigoyFRUVJzxmP369TOzZ8927z/++ONm4MCBZ62jQ4cO5tVXX/UYCwkJMS+99NIFdnR+fKXv7xo0aJC5//77L+g9F8JX+h4yZIj5+c9/3ug+LpSv9G2MMR9//LG5/PLLTWFhoZFk3nnnnUb1dD58qe/TvfHGG8bpdJqqqqoLet/58tW+P/nkEyPJHDx48ILed758re+NGzcaSeb48eMX3MuF8JW+Y2NjzcyZMxvfyAXgClEzeOONN9SnTx/17t1bEyZM0NKlS2X+9XVPOTk5iomJ8fhCyKSkJLlcLu3du7fB49XW1urEiRMKCQnxGD9w4IAiIyPVs2dPjR8/XgUFBR7zN954o15//XUdO3ZMtbW1WrFihcrLy5WQkNC0Df+Lr/R9ul27dik3N7dZ/8/Zl/rOyspS165ddfXVVysjI0OnTp1qwk49+Urfp06d0rhx47RkyZIGf5uwqflK399VWlqqoKAgtW3bPN+364t9l5WV6ZVXXlGPHj0UFRXVBF3W54t9twRf6Pvo0aPasWOHQkNDdeONNyosLExDhgw541Xii9YiscsyN954o1m0aJExxpiqqirTtWtXs3HjRmOMMZMnTzaJiYke68vKyowks3r16gaP9/TTT5vOnTub4uJi99jq1avNG2+8YT755BOzZs0aExcXZ7p3725cLpd7zfHjx01iYqKRZNq2bWuCgoLM+++/38Td/puv9H26qVOnmr59+zZBd2fmK33/9re/NWvWrDF5eXnmtddeM5dffrn5yU9+0sTd/puv9D1lyhSTkpLi3lczXyHylb5P9/XXX5vu3bubX/ziF03QYcN8qe8lS5aYDh06GEmmd+/ezXZ1yBjf6tuYlrtC5At95+TkGEkmJCTELF261Ozevds8/PDDxul0mi+++KLJeyYQNbHPP//ctG3b1uNfempqqpkwYYIx5sL/IGVlZZnAwECzbt26s573+PHjJigoyPzud79zj02bNs1cf/31Zv369SY3N9dkZmaa4OBgk5eXdzEtNsiX+q5z6tQpExwcbBYsWNCYls6LL/ZdZ8OGDc12K8FX+n733XfNVVddZU6cOOFe05yByFf6Pl1paam5/vrrzYgRI0xlZWVj2jonX+u7pKTEfPHFF2bz5s1m9OjR5tprrzXffvttY9s7I1/r25iWCUS+0vfWrVuNJJORkeGxLiYmxvzP//xPo3o7GwJRE5sxY4aRZNq0aePe/Pz8TEBAgCkpKTGPPfZYvXumX375pZFkdu/e7TH+hz/8wQQEBJhVq1ad17mvu+469x+SgwcPGklmz549HmuGDRtmHnjggcY3eAa+0vfpXn31VdOuXTtz9OjRRvd1Lr7Yd52TJ08aSWbNmjUX3Ne5+ErfP//5z43D4fCoQ5Lx8/MzQ4YMaYpWPfhK33VcLpeJi4szw4YNa5ZAUMfX+j5dRUWFCQwMNMuXL7/gvs7FF/tuiUDkK33XHfP//u//PNbcddddZty4cY1v8AwIRE2oqqrKhIWFmWeeecZ8+umnHluvXr3M888/734Y7fTk/dvf/tYEBQWZ8vJy99jy5ctN+/btzcqVK8/r3CdOnDCdO3c2ixcvNsYYk5eXZySZzz77zGNdYmKimTx5chN0+2++1PfphgwZYsaMGXPxDZ6Br/Zd569//auRZD755JPGN9kAX+q7sLCwXg2SzOLFi82XX355yfZtzD+vDN1www1myJAhpqysrOka/Q5f6/u7ysvLTUBAgHnllVca3WNDfLXv5g5EvtR3bW2tiYyMrPdQ9aBBg+pdNWoKBKIm9M477xin02lKSkrqzT366KPmuuuuM9XV1ebqq682iYmJJjc316xZs8Z069bN419uVlaWadu2rVmyZIkpLCx0b6cf95FHHjGbNm0y+fn5ZuvWrWb48OGma9eu7qshlZWV5qqrrjI333yz2bFjhzl48KBZsGCBcTgc5r333rtk+65z4MAB43A4zF/+8pcm7dVX+z548KCZM2eO2blzp8nPzzfvvvuu6dmzp4mPj7+k+25Ic90y86W+S0tLTWxsrImJiTEHDx70OE51dfUl2/ehQ4fMk08+aXbu3Gn+9re/ma1bt5rRo0ebkJAQj7+cL7W+jfln+P/444/NSy+9ZCSZ7Oxs8/HHH5tvvvnmku772WefNUFBQebNN980Bw4cMDNnzjTt27dvlkcBCERN6Ec/+pEZOXJkg3M7duxw/9/6V199ZZKTk01AQIDp2rWreeSRRzw+KjtkyBAjqd42adIk95q7777bREREGKfTaS6//HJz99131/sD8sUXX5g77rjDhIaGmsDAQDNgwIB6H8O/FPs2xpiMjAwTFRVlampqmrzfOr7Ud0FBgYmPjzchISHG39/fXHXVVWbGjBmmtLT0ku67Ic0ViHyp77qrBA1t+fn5l2zf//jHP0xycrIJDQ017dq1M1dccYUZN26c+fzzz5u0Z1/r25h/fkS9oeM09ZUxX+vbGGPmzp1rrrjiChMYGGji4uLMli1bmrTnOg5j/vU5OgAAAEvxPUQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAg6b777tPtt9/e5MfNzMzUoEGDmvy4AJoWgQiATztTUNm0aZMcDodKSkpavKY6y5YtU6dOnbx2fgBNh0AEAACsRyACcEn45ptvNHbsWF1++eUKDAxUTEyM/vCHP3iseeuttxQTE6OAgAB16dJFw4cPV1lZmceaBQsWKCIiQl26dFFqaqqqqqouqI6nnnpKYWFh6tixo1JSUlReXn7RvQFofgQiAJeE8vJyDR48WO+995727NmjKVOm6N5779WHH34oSSosLNTYsWN1//33a9++fdq0aZPuuOMOnf771hs3btShQ4e0ceNG/f73v9eyZcu0bNmy867hjTfeUGZmpp588knt3LlTERER+s1vftPUrQJoBvzaPQCfdt999+m1115T+/btPcZrampUXl6u48ePn/E5nh/96Efq06ePFixYoN27d2vw4MH66quvFB0d3eB5Nm3apEOHDqlNmzaSpLvuukt+fn5asWJFg8dftmyZHn74YfdzTDfeeKOuueYaLVmyxL3mhhtuUHl5uXJzcy+8eQAthitEAHze0KFDlZub67H97ne/81hTU1OjJ554QjExMQoJCdFll12m999/XwUFBZKkgQMHatiwYYqJidGdd96pl156ScePH/c4Rv/+/d1hSJIiIiJ09OjR865z3759io2N9RiLi4u70HYBeAGBCIDP69Chg6666iqP7fLLL/dYM3/+fC1evFjp6enauHGjcnNzlZSUpMrKSklSmzZttG7dOv3lL39Rv3799Nxzz6l3797Kz893H6Ndu3Yex3Q4HKqtrW3+BgF4HYEIwCVh69atuu222zRhwgQNHDhQPXv21BdffOGxxuFw6KabbtLs2bP18ccfy+l06p133mmyGvr27asdO3Z4jG3fvr3Jjg+g+bT1dgEA0BS+973v6a233tK2bdvUuXNnLVy4UMXFxerXr58kaceOHdqwYYMSExMVGhqqHTt26Ouvv1bfvn2brIaf//znuu+++3TdddfppptuUlZWlvbu3auePXs22TkANA8CEYBLwsyZM/Xll18qKSlJgYGBmjJlim6//XaVlpZKkoKCgpSdna1FixbJ5XIpOjpazzzzjJKTk5ushrvvvluHDh3So48+qvLyco0ZM0ZTp07V+++/32TnANA8+JQZAACwHs8QAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6/x/ZaRciz5ILGwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "count_dict = collections.Counter(hto_adata.obs['hash_id'])\n",
    "del count_dict['Doublet']\n",
    "temp_df = pd.DataFrame.from_dict(count_dict, orient = 'index').reset_index()\n",
    "temp_df.columns = ['Hash Id', 'count']\n",
    "\n",
    "sns.barplot(data = temp_df, x = 'Hash Id', y = 'count')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af96b957",
   "metadata": {},
   "source": [
    "Next, we cross reference with 10x filtered barcodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bea7e0d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_path = '/data/peer/landm/Projects/sail-dogma/yi-clone-samples/data/YF-2818_8HetClonesHashtagMix_hLLL_multiome/cr-arc-results/filtered_feature_bc_matrix.h5'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "77f56494",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/landm/anaconda3/envs/scanpy/lib/python3.8/site-packages/anndata/_core/anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/landm/anaconda3/envs/scanpy/lib/python3.8/site-packages/anndata/_core/anndata.py:1840: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read_10x_h5(adata_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "91eb49ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "cr_barcodes = adata.obs_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "49a6ef9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HTO Barcodes: 16150 barcodes.\n",
      "Shared Barcodes: 16150 barcodes.\n"
     ]
    }
   ],
   "source": [
    "print(f\"HTO Barcodes: {len(hto_barcodes)} barcodes.\")\n",
    "print(f\"Shared Barcodes: {len(set(hto_barcodes).intersection(set(cr_barcodes)))} barcodes.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cb8f0f9",
   "metadata": {},
   "source": [
    "All HTO barcodes are in 10x filtered barcodes. Next, we split and save each sample into a `barcodes.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7d6743b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "by_clone_cache_dir = f'/data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/{SAMPLE}/by_clone/'\n",
    "\n",
    "if not os.path.exists(f\"{by_clone_cache_dir}\"):\n",
    "    os.mkdir(by_clone_cache_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d2fa2f0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for hash_id in np.unique(hto_adata.obs['hash_id']):\n",
    "    if hash_id != 'Doublet':\n",
    "        new_dir = f\"{by_clone_cache_dir}{hash_id}/mito/\"\n",
    "        if not os.path.exists(f\"{new_dir}\"):\n",
    "            os.mkdir(new_dir)\n",
    "        \n",
    "        barcodes_df = pd.DataFrame(hto_adata[hto_adata.obs['hash_id'] == hash_id].obs['cr_barcodes'])\n",
    "        barcodes_df.to_csv(f\"{new_dir}barcodes.tsv\", index = None, header = None)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84be5aa3",
   "metadata": {},
   "source": [
    "# Process Mito Fragment Counts per Clone"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a789b097",
   "metadata": {},
   "source": [
    "# 00 Getting ATAC Fragments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c0de646",
   "metadata": {},
   "source": [
    "**NOTE:** We obtained our mito fragments from the fastq files by running Cellranger ARC for multiomic data. \n",
    "\n",
    "**However, we made the following changes to run Cellranger ARC:**\n",
    "- Used a custom CellRanger ARC reference genome. We blacklisted nuclear regions that were homologous to the mitochondrial genome in our reference genome. We followed the steps here (https://github.com/caleblareau/mitoblacklist). This prevents mito reads from being multimapped and thrown out and increased the coverage of our mitochondrial reads.\n",
    "- Move `chrM` from `non_nuclear_contigs` to `primary_contigs` in the CellRanger ARC configuration file. This ensures the mitochondrial reads are not discarded during processing.\n",
    "\n",
    "After processing, CellRanger ARC provides us with `atac_fragments.tsv.gz` as output. This file contains information on all the ATAC fragments sequenced. This file is used as input for the next steps of processing:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcfdb59a",
   "metadata": {},
   "source": [
    "Step 01 subsets the CR-Arc fragments for Mito fragments. We do this once for all barcodes.\n",
    "\n",
    "Step 02 and 03 are then done for **each clone** (by subsetting using the barcodes above)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "49f6f249",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragments_path = f'/data/peer/landm/Projects/sail-dogma/yi-clone-samples/data/{SAMPLE}/cr-arc-results/atac_fragments.tsv.gz'\n",
    "cache_dir = f'/data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/'\n",
    "by_clone_cache_dir = f'/data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/{SAMPLE}/by_clone/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04e0763b",
   "metadata": {},
   "source": [
    "# 01 Subset Multiome ATAC fragments to just Mito"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a65190b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragments = pd.read_csv(f\"{fragments_path}\", compression='gzip', header = None, sep='\\t', quotechar='\"', comment = '#')\n",
    "mito_fragments = fragments[fragments[0] == 'chrM']\n",
    "print(f'{mito_fragments.shape[0]} mitochondrial fragments')\n",
    "\n",
    "# Annotate fragments by where they fall on the genome\n",
    "# TODO: This doesn't take into account a fragment that can span multiple genes\n",
    "mito_fragments['gene'] = 'chrM'\n",
    "mito_fragments.columns = ['chromosome', 'start', 'end', 'barcode', 'umis', 'gene']\n",
    "for row in mito_genes.iterrows():\n",
    "    row = row[1]\n",
    "    start = row['start']\n",
    "    end = row['end']\n",
    "\n",
    "    mito_fragments.loc[(mito_fragments['start']>=start) & (mito_fragments['end'] < end), 'gene'] = row['gene_name']\n",
    "# mito_fragments.head()\n",
    "\n",
    "mito_dir = f\"{cache_dir}{SAMPLE}/mito/\"\n",
    "if not os.path.exists(mito_dir):\n",
    "    os.mkdir(mito_dir)\n",
    "mito_fragments.to_csv(f'{mito_dir}mito_fragments.csv')\n",
    "\n",
    "# Get counts for each gene per cell barcode\n",
    "mito_counts = mito_fragments[['barcode', 'gene']].groupby('barcode').value_counts().unstack(fill_value = 0)\n",
    "mito_counts.head()\n",
    "mito_counts.to_csv(f'{mito_dir}mito_fragments_counts.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26617bcb",
   "metadata": {},
   "source": [
    "## Optional: Filter each into separate dfs for each clone."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c919a82",
   "metadata": {},
   "source": [
    "In our case, we separated the mito fragments file by clone to make the next steps computationally faster. Otherwise we would first have to filter the entire mito fragments file for each clone."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a7caae05",
   "metadata": {},
   "outputs": [],
   "source": [
    "mito_frag_path = f'{cache_dir}{SAMPLE}/mito/mito_fragments.csv'\n",
    "\n",
    "df = pd.read_csv(f'{mito_dir}mito_fragments.csv', index_col = 0) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a13dcd73",
   "metadata": {},
   "outputs": [],
   "source": [
    "for clone in os.listdir(by_clone_cache_dir):\n",
    "    barcodes_path = f\"{by_clone_cache_dir}{clone}/mito/barcodes.tsv\"\n",
    "    barcodes = pd.read_csv(barcodes_path, index_col = None, header = None)\n",
    "    out_path = f\"{by_clone_cache_dir}{clone}/mito/mito_fragments.csv\"\n",
    "    # df[df['barcode'].isin(barcodes.values.flatten())].to_csv(out_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ef588c9",
   "metadata": {},
   "source": [
    "## Split frags greater than 3000 bp\n",
    "\n",
    "From other analysis, we know that frags greater than 3000 bp mostly span the deleted region and are therefore representative of endjoined mito. Thus, we split the fragments into two with a padding of 50. \n",
    "\n",
    "For example,\n",
    "\n",
    "Fragment `7692 - 11869` --> `7692 - 7742` and `11819 - 11869`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b0eb11c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_long_frag_df(df, frag_threshold, padding):\n",
    "    assert padding >= 0, \"padding must be positive\"\n",
    "    \n",
    "    long_frags_df = df[(df['end'] - df['start']) >= frag_threshold]\n",
    "    start_frags_df =  long_frags_df.copy()\n",
    "    start_frags_df['end'] = start_frags_df['start'] + padding\n",
    "    end_frags_df =  long_frags_df.copy()\n",
    "    end_frags_df['start'] = end_frags_df['end'] - padding\n",
    "    \n",
    "    \n",
    "    return pd.concat([start_frags_df, end_frags_df]).sort_values('start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "dbe4ab97",
   "metadata": {},
   "outputs": [],
   "source": [
    "frag_threshold = 3000\n",
    "padding = 50\n",
    "\n",
    "for clone in os.listdir(by_clone_cache_dir):\n",
    "    unfilt_path = f\"{by_clone_cache_dir}{clone}/mito/mito_fragments.csv\"\n",
    "    filt_path = f\"{by_clone_cache_dir}{clone}/mito/mito_fragments_split_{frag_threshold}_bp_frags.csv\"\n",
    "    \n",
    "    df = pd.read_csv(unfilt_path, index_col = 0)\n",
    "    no_long_frags_df = df[(df['end'] - df['start']) < frag_threshold]\n",
    "    split_frags_df = split_long_frag_df(df, frag_threshold, padding)\n",
    "    new_frags_df = pd.concat([no_long_frags_df, split_frags_df]).sort_values('start')\n",
    "    \n",
    "    # new_frags_df.to_csv(filt_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "6cf50e41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>barcode</th>\n",
       "      <th>umis</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>203821859</th>\n",
       "      <td>chrM</td>\n",
       "      <td>4</td>\n",
       "      <td>24</td>\n",
       "      <td>CTGATCACACTTACAG-1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203833529</th>\n",
       "      <td>chrM</td>\n",
       "      <td>4</td>\n",
       "      <td>105</td>\n",
       "      <td>GCAGCCAGTATTGGAT-1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203833526</th>\n",
       "      <td>chrM</td>\n",
       "      <td>4</td>\n",
       "      <td>105</td>\n",
       "      <td>GCAACAGCAATTGCGC-1</td>\n",
       "      <td>2</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203833460</th>\n",
       "      <td>chrM</td>\n",
       "      <td>4</td>\n",
       "      <td>105</td>\n",
       "      <td>CATGAGGCAGTCTATG-1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203833457</th>\n",
       "      <td>chrM</td>\n",
       "      <td>4</td>\n",
       "      <td>105</td>\n",
       "      <td>CATAATCCAATCATGT-1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289860924</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16553</td>\n",
       "      <td>16569</td>\n",
       "      <td>TTCCACGGTCATTGTC-1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289860880</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16553</td>\n",
       "      <td>16569</td>\n",
       "      <td>AATATCCCACCCACAG-1</td>\n",
       "      <td>2</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289860898</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16553</td>\n",
       "      <td>16569</td>\n",
       "      <td>CGCTGTGCAGCAACCT-1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289860899</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16553</td>\n",
       "      <td>16569</td>\n",
       "      <td>CGGTTATAGTCATTTC-1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289860913</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16553</td>\n",
       "      <td>16569</td>\n",
       "      <td>GCGTGCTAGCAGCTCA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6794840 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chromosome  start    end             barcode  umis  gene\n",
       "203821859       chrM      4     24  CTGATCACACTTACAG-1     1  chrM\n",
       "203833529       chrM      4    105  GCAGCCAGTATTGGAT-1     1  chrM\n",
       "203833526       chrM      4    105  GCAACAGCAATTGCGC-1     2  chrM\n",
       "203833460       chrM      4    105  CATGAGGCAGTCTATG-1     1  chrM\n",
       "203833457       chrM      4    105  CATAATCCAATCATGT-1     1  chrM\n",
       "...              ...    ...    ...                 ...   ...   ...\n",
       "289860924       chrM  16553  16569  TTCCACGGTCATTGTC-1     1  chrM\n",
       "289860880       chrM  16553  16569  AATATCCCACCCACAG-1     2  chrM\n",
       "289860898       chrM  16553  16569  CGCTGTGCAGCAACCT-1     1  chrM\n",
       "289860899       chrM  16553  16569  CGGTTATAGTCATTTC-1     1  chrM\n",
       "289860913       chrM  16553  16569  GCGTGCTAGCAGCTCA-1     1  chrM\n",
       "\n",
       "[6794840 rows x 6 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_frags_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7aee0f0",
   "metadata": {},
   "source": [
    "# 02 Generate Bulk Mitochondrial Coverage Plots\n",
    "\n",
    "This step just gives you an initial visualization of your mito coverage in bulk. You may choose to skip and generate coverage at a single-cell level, and then plot those results. \n",
    "\n",
    "Note you can also obtain these bulk results coverage after generating single-cell coverage results-- the reason this is a separate step is that assigning reads at single-cell level is slow. So if you just want mito coverage in bulk, this is a way to do it without running step 03."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a12b423",
   "metadata": {},
   "source": [
    "First, for each clone, we count the number of fragments that overlap each basepair from 0-16569 and save it as a dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "090f91f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job <4908470> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0257_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0257.stdout -eo A0257.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/make_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --mito_fragments_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0257/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0257/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0257/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
      "\n",
      "Job <4908472> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0256_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0256.stdout -eo A0256.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/make_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --mito_fragments_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0256/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0256/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0256/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
      "\n",
      "Job <4908473> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0253_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0253.stdout -eo A0253.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/make_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --mito_fragments_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0253/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0253/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0253/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
      "\n",
      "Job <4908474> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0258_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0258.stdout -eo A0258.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/make_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --mito_fragments_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0258/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0258/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0258/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
      "\n",
      "Job <4908475> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0255_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0255.stdout -eo A0255.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/make_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --mito_fragments_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0255/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0255/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0255/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
      "\n",
      "Job <4908476> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0252_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0252.stdout -eo A0252.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/make_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --mito_fragments_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0252/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0252/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0252/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
      "\n",
      "Job <4908477> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0254_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0254.stdout -eo A0254.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/make_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --mito_fragments_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0254/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0254/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0254/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
      "\n",
      "Job <4908478> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0251_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0251.stdout -eo A0251.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/make_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --mito_fragments_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0251/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0251/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0251/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# relative to repo\n",
    "cov_script = \"../../scripts/make_coverage.py\"\n",
    "mito_genes_path = '../../misc/mito_genes.csv'\n",
    "\n",
    "for clone in os.listdir(by_clone_cache_dir):\n",
    "    # input path to mito fragments \n",
    "    mito_frag_path = f'{by_clone_cache_dir}{clone}/mito/mito_fragments_split_3000_bp_frags.csv'\n",
    "    # input path to cell barcodes for sample\n",
    "    barcodes_path = f\"{by_clone_cache_dir}{clone}/mito/barcodes.tsv\"\n",
    "    # output path to store bulk coverage csv\n",
    "    out_path = f\"{by_clone_cache_dir}{clone}/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
    "    memory = 10\n",
    "    conda_env = 'scanpy'\n",
    "    options = dict()\n",
    "    options['J'] = f'{clone}_coverage'\n",
    "    options['n'] = f'5'\n",
    "    options['R'] = f'rusage[mem={memory}]'\n",
    "    options['W'] = f'2:00'\n",
    "    options['o'] = f'{clone}.stdout'\n",
    "    options['eo'] = f'{clone}.stdout'\n",
    "\n",
    "\n",
    "    if os.path.exists(options['o']):\n",
    "        os.remove(options['o'])\n",
    "\n",
    "    option_string = ' '.join([f'-{flag} {val}' for flag, val in options.items()])\n",
    "    job_string = f\"source ~/.bashrc; conda activate {conda_env};\"\n",
    "    \n",
    "    # THIS IS THE ACTUAL COMMAND LINE TO RUN THE SCRIPT (everything else is to allow it to run on MSKCC servers)\n",
    "    run_script_string = f\"python {cov_script} --mito_genes_path {mito_genes_path} --mito_fragments_path {mito_frag_path}\"\n",
    "    script2 = f\"--barcodes_path {barcodes_path} --out_path {out_path}\"\n",
    "\n",
    "    command = f'bsub {option_string} \\\"{job_string} {run_script_string} {script2}\"'\n",
    "    os.system(command)\n",
    "    print(command + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52fd299b",
   "metadata": {},
   "source": [
    "The following code takes the dataframe generated above and plots it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "76d691b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Job <4908670> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0257_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0257.stdout -eo A0257.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/plot_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --cov_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0257/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0257/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0257/mito/mito_split_3000_bp_frags_coverage_filt_plot_normalized.png --sample A0257\"\n",
      "\n",
      "Job <4908671> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0256_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0256.stdout -eo A0256.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/plot_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --cov_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0256/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0256/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0256/mito/mito_split_3000_bp_frags_coverage_filt_plot_normalized.png --sample A0256\"\n",
      "\n",
      "Job <4908672> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0253_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0253.stdout -eo A0253.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/plot_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --cov_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0253/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0253/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0253/mito/mito_split_3000_bp_frags_coverage_filt_plot_normalized.png --sample A0253\"\n",
      "\n",
      "Job <4908673> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0258_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0258.stdout -eo A0258.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/plot_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --cov_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0258/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0258/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0258/mito/mito_split_3000_bp_frags_coverage_filt_plot_normalized.png --sample A0258\"\n",
      "\n",
      "Job <4908674> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0255_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0255.stdout -eo A0255.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/plot_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --cov_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0255/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0255/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0255/mito/mito_split_3000_bp_frags_coverage_filt_plot_normalized.png --sample A0255\"\n",
      "\n",
      "Job <4908675> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0252_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0252.stdout -eo A0252.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/plot_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --cov_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0252/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0252/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0252/mito/mito_split_3000_bp_frags_coverage_filt_plot_normalized.png --sample A0252\"\n",
      "\n",
      "Job <4908676> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0254_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0254.stdout -eo A0254.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/plot_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --cov_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0254/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0254/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0254/mito/mito_split_3000_bp_frags_coverage_filt_plot_normalized.png --sample A0254\"\n",
      "\n",
      "Job <4908677> is submitted to default queue <cpuqueue>.\n",
      "bsub -J A0251_coverage -n 5 -R rusage[mem=10] -W 2:00 -o A0251.stdout -eo A0251.stdout \"source ~/.bashrc; conda activate scanpy; python /data/peer/landm/Projects/sail-dogma/scripts/coverage/plot_coverage.py --mito_genes_path /data/peer/landm/Projects/sail-dogma/4-cell-line/cache/mito_processing_cache/mito_genes.csv --cov_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0251/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0251/mito/barcodes.tsv --out_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0251/mito/mito_split_3000_bp_frags_coverage_filt_plot_normalized.png --sample A0251\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# relative to repo\n",
    "plot_script = \"../../scripts/plot_coverage.py\"\n",
    "mito_genes_path = '../../misc/mito_genes.csv'\n",
    "\n",
    "for clone in os.listdir(by_clone_cache_dir):\n",
    "    # input cell barcodes for sample\n",
    "    barcodes_path = f\"{by_clone_cache_dir}{clone}/mito/barcodes.tsv\"\n",
    "    # input bulk coverage csv (output of above code)\n",
    "    cov_path = f\"{by_clone_cache_dir}{clone}/mito/mito_fragments_split_3000_bp_frags_coverage_filt.csv\"\n",
    "    # output path to store result\n",
    "    out_path = f\"{by_clone_cache_dir}{clone}/mito/mito_split_3000_bp_frags_coverage_filt_plot_normalized.png\"\n",
    "    memory = 10\n",
    "    conda_env = 'scanpy'\n",
    "    options = dict()\n",
    "    options['J'] = f'{clone}_coverage'\n",
    "    options['n'] = f'5'\n",
    "    options['R'] = f'rusage[mem={memory}]'\n",
    "    options['W'] = f'2:00'\n",
    "    options['o'] = f'{clone}.stdout'\n",
    "    options['eo'] = f'{clone}.stdout'\n",
    "\n",
    "\n",
    "    if os.path.exists(options['o']):\n",
    "        os.remove(options['o'])\n",
    "\n",
    "    option_string = ' '.join([f'-{flag} {val}' for flag, val in options.items()])\n",
    "    job_string = f\"source ~/.bashrc; conda activate {conda_env};\"\n",
    "    \n",
    "    # THIS IS THE ACTUAL COMMAND LINE TO RUN THE SCRIPT (everything else is to allow it to run on MSKCC servers)\n",
    "    run_script_string = f\"python {plot_script} --mito_genes_path {mito_genes_path} --cov_path {cov_path}\"\n",
    "    script2 = f\"--barcodes_path {barcodes_path} --out_path {out_path} --sample {clone}\"\n",
    "\n",
    "    command = f'bsub {option_string} \\\"{job_string} {run_script_string} {script2}\"'\n",
    "    os.system(command)\n",
    "    print(command + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6bf3fa8",
   "metadata": {},
   "source": [
    "# 03 Compute Single-cell coverage (takes longer to run)\n",
    "\n",
    "For each cell, we report the number of unique fragments (different START and END) that overlap each base pair."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3c36e3fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bsub -J A0257_coverage -n 5 -R rusage[mem=10] -W 120:00 -o A0257.stdout -eo A0257.stdout \"source ~/.bashrc; conda activate scanpy; python ../../scripts/single-cell-coverage.py --mito_genes_path ../../misc/mito_genes.csv --mito_frags_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0257/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0257/mito/barcodes.tsv --out_dir /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0257/mito/single-cell-coverage_split_3000_bp_frags/\"\n",
      "\n",
      "bsub -J A0256_coverage -n 5 -R rusage[mem=10] -W 120:00 -o A0256.stdout -eo A0256.stdout \"source ~/.bashrc; conda activate scanpy; python ../../scripts/single-cell-coverage.py --mito_genes_path ../../misc/mito_genes.csv --mito_frags_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0256/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0256/mito/barcodes.tsv --out_dir /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0256/mito/single-cell-coverage_split_3000_bp_frags/\"\n",
      "\n",
      "bsub -J A0253_coverage -n 5 -R rusage[mem=10] -W 120:00 -o A0253.stdout -eo A0253.stdout \"source ~/.bashrc; conda activate scanpy; python ../../scripts/single-cell-coverage.py --mito_genes_path ../../misc/mito_genes.csv --mito_frags_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0253/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0253/mito/barcodes.tsv --out_dir /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0253/mito/single-cell-coverage_split_3000_bp_frags/\"\n",
      "\n",
      "bsub -J A0258_coverage -n 5 -R rusage[mem=10] -W 120:00 -o A0258.stdout -eo A0258.stdout \"source ~/.bashrc; conda activate scanpy; python ../../scripts/single-cell-coverage.py --mito_genes_path ../../misc/mito_genes.csv --mito_frags_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0258/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0258/mito/barcodes.tsv --out_dir /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0258/mito/single-cell-coverage_split_3000_bp_frags/\"\n",
      "\n",
      "bsub -J A0255_coverage -n 5 -R rusage[mem=10] -W 120:00 -o A0255.stdout -eo A0255.stdout \"source ~/.bashrc; conda activate scanpy; python ../../scripts/single-cell-coverage.py --mito_genes_path ../../misc/mito_genes.csv --mito_frags_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0255/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0255/mito/barcodes.tsv --out_dir /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0255/mito/single-cell-coverage_split_3000_bp_frags/\"\n",
      "\n",
      "bsub -J A0252_coverage -n 5 -R rusage[mem=10] -W 120:00 -o A0252.stdout -eo A0252.stdout \"source ~/.bashrc; conda activate scanpy; python ../../scripts/single-cell-coverage.py --mito_genes_path ../../misc/mito_genes.csv --mito_frags_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0252/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0252/mito/barcodes.tsv --out_dir /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0252/mito/single-cell-coverage_split_3000_bp_frags/\"\n",
      "\n",
      "bsub -J A0254_coverage -n 5 -R rusage[mem=10] -W 120:00 -o A0254.stdout -eo A0254.stdout \"source ~/.bashrc; conda activate scanpy; python ../../scripts/single-cell-coverage.py --mito_genes_path ../../misc/mito_genes.csv --mito_frags_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0254/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0254/mito/barcodes.tsv --out_dir /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0254/mito/single-cell-coverage_split_3000_bp_frags/\"\n",
      "\n",
      "bsub -J A0251_coverage -n 5 -R rusage[mem=10] -W 120:00 -o A0251.stdout -eo A0251.stdout \"source ~/.bashrc; conda activate scanpy; python ../../scripts/single-cell-coverage.py --mito_genes_path ../../misc/mito_genes.csv --mito_frags_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0251/mito/mito_fragments_split_3000_bp_frags.csv --barcodes_path /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0251/mito/barcodes.tsv --out_dir /data/peer/landm/Projects/sail-dogma/yi-clone-samples/cache/YF-2818_8HetClonesHashtagMix_hLLL_multiome/by_clone/A0251/mito/single-cell-coverage_split_3000_bp_frags/\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# relative to repo\n",
    "cov_script = \"../../scripts/single-cell-coverage.py\"\n",
    "mito_genes_path = '../../misc/mito_genes.csv'\n",
    "\n",
    "for clone in os.listdir(by_clone_cache_dir):\n",
    "    mito_frag_path = f'{by_clone_cache_dir}{clone}/mito/mito_fragments_split_3000_bp_frags.csv'\n",
    "    barcodes_path = f\"{by_clone_cache_dir}{clone}/mito/barcodes.tsv\"\n",
    "    out_dir = f'{by_clone_cache_dir}{clone}/mito/single-cell-coverage_split_3000_bp_frags/'\n",
    "    if not os.path.exists(out_dir):\n",
    "        os.mkdir(out_dir)\n",
    "    \n",
    "    memory = 10\n",
    "    conda_env = 'scanpy'\n",
    "    options = dict()\n",
    "    options['J'] = f'{clone}_coverage'\n",
    "    options['n'] = f'5'\n",
    "    options['R'] = f'rusage[mem={memory}]'\n",
    "    options['W'] = f'120:00'\n",
    "    options['o'] = f'{clone}.stdout'\n",
    "    options['eo'] = f'{clone}.stdout'\n",
    "\n",
    "\n",
    "    if os.path.exists(options['o']):\n",
    "        os.remove(options['o'])\n",
    "\n",
    "    option_string = ' '.join([f'-{flag} {val}' for flag, val in options.items()])\n",
    "    job_string = f\"source ~/.bashrc; conda activate {conda_env};\"\n",
    "    \n",
    "    # THIS IS THE ACTUAL COMMAND LINE TO RUN THE SCRIPT (everything else is to allow it to run on MSKCC servers)\n",
    "    run_script_string = f\"python {cov_script} --mito_genes_path {mito_genes_path} --mito_frags_path {mito_frag_path} --barcodes_path {barcodes_path} --out_dir {out_dir}\"\n",
    "\n",
    "    command = f'bsub {option_string} \\\"{job_string} {run_script_string}\"'\n",
    "    # os.system(command)\n",
    "    print(command + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98e4b75d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
